src/basic/process-util.c

```diff
@@ -2054,14 +2060,55 @@ int posix_spawn_wrapper(const char *path, char *const *argv, char *const *envp,
         r = posix_spawnattr_init(&attr);
         if (r != 0)
                 return -r; /* These functions return a positive errno on failure */
-        /* Set all signals to SIG_DFL */
-        r = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGMASK|POSIX_SPAWN_SETSIGDEF);
+
+#if HAVE_PIDFD_SPAWN
+        if (cgroup) {
+                r = cg_get_path_and_check(
+                                SYSTEMD_CGROUP_CONTROLLER,
+                                cgroup,
+                                /* suffix= */ NULL,
+                                &resolved_cgroup);
+                if (r < 0)
+                        goto fail;
+
+                cgroup_fd = open(resolved_cgroup, O_PATH|O_DIRECTORY|O_CLOEXEC);
+                if (cgroup_fd < 0) {
+                        r = errno;
+                        goto fail;
+                }
+
+                r = posix_spawnattr_setcgroup_np(&attr, cgroup_fd);
+                if (r != 0)
+                        goto fail;
+
+                flags |= POSIX_SPAWN_SETCGROUP;
+        }
+#endif
+
+        r = posix_spawnattr_setflags(&attr, flags);
         if (r != 0)
                 goto fail;
         r = posix_spawnattr_setsigmask(&attr, &mask);
         if (r != 0)
                 goto fail;
 
+#if HAVE_PIDFD_SPAWN
+        r = pidfd_spawn(&pidfd, path, NULL, &attr, argv, envp);
+        if (r == 0) {
+                posix_spawnattr_destroy(&attr);
```

Comment:
hmm, let's just make that a regular clean up thing, i.e. drop this, and the call at the end, and just define the local variable as `_cleanup_(posix_spawnattr_destroy) posix_spawnattr_t attr;`. That should suffix and make the error path here a lot simpler.

Reply:
eh unfortunately that gets messy because of the very first error check. If that fails, the attr is not initialized, and I can't remember if it's the compiler or the static analyzer or both that freak out. I tried that first when I first added this wrapper. Even asserting instead of error checking was not enough to make it work.

Reply:
yuck. maybe just use DEFINE_TRIVIAL_CLEANUP() to turn posix_spawnattr_destroy() into a destructor that takes a pointer, and then add a definition like this to the right place:

```c
_cleanup_(posix_spawnattr_destroyp) posix_spawn_attr_t *attr_destructor = &attr;
```

from which point on auto-destruction will be enabled.

Reply:
Ah nice that seems to work - let's see if the CI agrees